model/common/src/icon4py/model/common/grid/vertical.py

```diff
@@ -517,3 +539,141 @@ def get_vct_a_and_vct_b(vertical_config: VerticalGridConfig) -> tuple[fa.KField,
         if vertical_config.file_path
         else _compute_vct_a_and_vct_b(vertical_config)
     )
+
+
+def init_vert_coord(
+    vct_a: fa.KField[ta.wpfloat],
+    topography: fa.CellField[ta.wpfloat],
+    topography_smoothed: fa.CellField[ta.wpfloat],
+    grid: icon_grid.IconGrid,
+    vertical_config: VerticalGridConfig,
+    vertical_geometry: VerticalGrid,
+    nshift: int = 0,  # shift parameter used for vertical nesting
+) -> fa.CellKField[ta.wpfloat]:
+    """
+    Computes the 3D vertical coordinate fields.
+    """
+
+    def _decay_func(vct_a, model_top_height, decay_scale, decay_exponent):
+        return xp.sinh(
+            (model_top_height / decay_scale) ** decay_exponent
+            - (vct_a / decay_scale) ** decay_exponent
+        ) / xp.sinh((model_top_height / decay_scale) ** decay_exponent)
+
+    z3d_i = xp.zeros((grid.num_cells, grid.num_levels + 1), dtype=ta.wpfloat)
+
+    topography = topography.asnumpy()
+    topography_smoothed = topography_smoothed.asnumpy()
+    vct_a = vct_a.asnumpy()
+
+    z3d_i[:, grid.num_levels] = topography
+    ktop_thicklimit = xp.asarray(grid.num_cells * [grid.num_levels], dtype=ta.wpfloat)
+
+    # Small-scale topography (i.e. full topo - smooth topo)
+    topo_deviation = topography - topography_smoothed
+
+    k = range(vertical_geometry.nflatlev + 1)
+    k1 = range(nshift, vertical_geometry.nflatlev + 1 + nshift)
+    z3d_i[:, k] = vct_a[k1]
+
+    k = range(vertical_geometry.nflatlev + 1, grid.num_levels)
+    k1 = range(vertical_geometry.nflatlev + 1 + nshift, grid.num_levels + nshift)
+    # Scaling factors for large-scale and small-scale topography
+    z_fac1 = _decay_func(
+        vct_a[k1],
+        vertical_config.model_top_height,
+        vertical_config.decay_scale_1,
+        vertical_config.decay_exponent,
+    )
+    z_fac2 = _decay_func(
+        vct_a[k1],
+        vertical_config.model_top_height,
+        vertical_config.decay_scale_2,
+        vertical_config.decay_exponent,
+    )
+    z3d_i[:, k] = (
+        vct_a[k1][xp.newaxis, :]
+        + topography_smoothed[:, xp.newaxis] * z_fac1
+        + topo_deviation[:, xp.newaxis] * z_fac2
+    )
+
+    # Ensure that layer thicknesses are not too small; this would potentially
+    # cause instabilities in vertical advection
+    for k in reversed(range(grid.num_levels)):
+        k1 = k + nshift
+        dvct = vct_a[k1] - vct_a[k1 + 1]
```

Comment:
could you come up with better names for these variables? Especially the non local ones, for example the ones in `vertical_config` ?

Reply:
I had stuck to fortran names, now updated, hopefully better